### PR TITLE
[fix]: [CDS-71025]: Consider TaskSelector in prepareTaskRequest Steps

### PR DESCRIPTION
**Pull request by Abhishek-CDC**
## Describe your changes
When delegate selectors are passed as list of string the origin information is lost. Because of this, delegate selectors of both connector and step are appearing.
Made changes to consider TaskSelector in place of string for creating TaskRequest.
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50724)

### CODE REVIEW

**Comment by Abhishek-CDC** [comment]
trigger uts

**Comment by Abhishek-CDC** [comment]
trigger uts

File: pipeline-service/service/src/main/java/io/harness/pms/jira/JiraStepHelperServiceImpl.java [hunk 1]
```diff
@@ -70,7 +72,7 @@ public JiraStepHelperServiceImpl(ConnectorResourceClient connectorResourceClient
 
   @Override
   public TaskRequest prepareTaskRequest(JiraTaskNGParametersBuilder paramsBuilder, Ambiance ambiance,
-      String connectorRef, String timeStr, String taskName) {
+      String connectorRef, String timeStr, String taskName, SpecParameters specParameters) {
```
Comment (yogesh-chauhan) [inline pipeline-service/service/src/main/java/io/harness/pms/jira/JiraStepHelperServiceImpl.java:75]: what is we pass taskSelectors directly ?
will it help getting rid of the instanceOf check in this method ?

File: pipeline-service/service/src/main/java/io/harness/pms/servicenow/ServiceNowStepHelperServiceImpl.java [hunk 2]
```diff
@@ -81,7 +84,7 @@ public ServiceNowStepHelperServiceImpl(ConnectorResourceClient connectorResource
 
   @Override
   public TaskRequest prepareTaskRequest(ServiceNowTaskNGParametersBuilder paramsBuilder, Ambiance ambiance,
-      String connectorRef, String timeStr, String taskName) {
+      String connectorRef, String timeStr, String taskName, SpecParameters specParameters) {
```
Comment (yogesh-chauhan) [inline pipeline-service/service/src/main/java/io/harness/pms/servicenow/ServiceNowStepHelperServiceImpl.java:87]: same comment as above

**Review by yogesh-chauhan — APPROVED** [review]

**Review by puthrayaharness — DISMISSED** [review]

**Comment by Abhishek-CDC** [comment]
trigger uts

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/bamboo/BambooBuildStepHelperService.java` line 27 at r4](https://reviewable.io/reviews/harness/harness-core/50724#-NaFiDaw7T-85F5pMbep:-NaFiDaw7T-85F5pMbeq:bmsgvau) ([raw file](https://github.com/harness/harness-core/blob/45ab8faefcf16f44963e0622f0ec8aacd5e3e2d0/125-cd-nextgen/src/main/java/io/harness/cdng/bamboo/BambooBuildStepHelperService.java#L27)):*
> ```Java
> public interface BambooBuildStepHelperService {
>   TaskRequest prepareTaskRequest(BambooArtifactDelegateRequestBuilder paramsBuilder, Ambiance ambiance,
>       String connectorRef, String timeStr, String taskName, ParameterField<List<TaskSelectorYaml>> delegateSelectors);
> ```

we should not pass parameter field. We should pass List<TaskSelectorYaml>

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 24 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50724)* status: 30 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by Abhishek-CDC** [comment]
trigger uts

**Comment by Abhishek-CDC** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 5 files at r2, 4 of 24 files at r3, 16 of 16 files at r5, 2 of 2 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50724)* status: 30 file(s) to review. 1 of 2 approvals obtained.

**Review by puthrayaharness — APPROVED** [review]

**Comment by puthrayaharness** [comment]
trigger review

**Comment by Abhishek-CDC** [comment]
trigger messagecheck

**Comment by Abhishek-CDC** [comment]
trigger messagecheck

**Pull request merged**